### PR TITLE
[GEOT-6900] Shapefile quadtree build performance improvements

### PR DESCRIPTION
**Pull request by groldan**
[GEOT-6900](https://osgeo-org.atlassian.net/browse/GEOT-6900) [Powered by Pull Request Badge](https://pullrequestbadge.com/?utm_medium=github&utm_source=geotools&utm_campaign=badge_info)

Use a strategy object (`BoundsReader`) to assist `ShapeFileIndexer` in speeding up the 
`QuadTree` optimization  phase, providing quick access to each shapefile
 record envelope, potentially avoiding an immense amount of random disk I/O calls through
`ShapefileReader`, as the quad tree internal nodes get split/shrank.

Since the `QuadTree` leaf nodes hold only the shapefile record ids, and not their
 bounds, the tree layout optimization phase may incur into too much random disk reads on the
`.shp` file, which has a bigger impact the bigger the shapefile is, especially related to the size of the
geometries more than the number of records itself.

The `BoundsReader` strategy object is meant to avoid that to the extent possible.

To a given point, record bounds will be stored in heap memory (up to 1MiB, accounting for 32K
 records, or 64K records if it's a points shapefile).

For a bigger number of shapefile records, the strategy is
 to store the bounds in a temporary file (named 
 `GeoTools_shp_qix_bounds_<random number>.tmp`  under `${java.io.tmpdir}`), 
 which is memory mapped and deleted at `BoundsReader.close()`. This leverages the Operating 
 System's native paging, and due to the reduced size of the bounds file compared to the actual 
 `.shp` and avoiding the parsing performed by `ShapefileReader.nextRecord()`, results in 
 dramatically less random I/O and computing.

Note, however, that if there's not enough temporary space in the file system where the 
`java.io.tmpdir` directory resides, a fall back strategy that reads directly from the 
`ShapefileReader` will be used. This should be a very edge case though, since with a bounds record
 size of 32 bytes, the required temporary storage is 30.1MiB per million features.

## Index build time

The polygon test dataset that was used is attached to [GEOT-6900](https://osgeo-org.atlassian.net/browse/GEOT-6900).

Running `ShapefileIndexer` on it from `main` at 0adffaf3278ca8ff7f6e653b33749620f6cbf59c, prints:
```
341769 features indexed in 193826ms.
```
The same execution with this patch prints:
```
341769 features indexed in 920ms.
```
and produces the exact same results.

The new `ShapefileIndexerStressTest.java` test, ran against the above-mentioned baseline and this patch, results in:

|# Features| Baseline | This patch |
|------------------|----------------| ----------------|
|1,000            | 15 ms| 16 ms|
|10,000         | 41 ms| 24 ms|
|100,000       | 204 ms| 106 ms|
|1,000,000   | 1821 ms| 600 ms|
|10,000,000| 21994 s| 5720 ms|

This shows that the impact of the improvement is proportional to the number of records, but more importantly to the complexity/size of the data itself. Since the test data produced by the stress test is very simple, with 10M records we get only a 4x improvement, but with 300K complex polygons the improvement is 2 orders of magnitude.

# Checklist

Reviewing is a process done by project maintainers, **mostly on a volunteer basis** (thus limited in time). We need to keep the review overhead as small as possible, and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

For all pull requests:

- [x] Confirm you have read the [contribution guidelines](https://github.com/geotools/geotools/blob/main/CONTRIBUTING.md)
- [x] You have sent a Contribution Licence Agreement (CLA) as necessary (not required for small changes, e.g., fixing typos in documentation)
- [x] Make sure the first PR targets the `main` branch, eventual backports will be managed later. This can be ignored if the PR is fixing an issue that only happens in a specific branch, but not in newer ones.
- [x] The changes are not causing two modules to share the same Java packages (to avoid [Java 9+ split package](http://tutorials.jenkov.com/java/modules.html#split-packages-not-allowed) issues)
- [x] The changes are not breaking the build in downstream projects using SNAPSHOT dependencies, GeoWebCache and GeoServer (there is an automatic PR check verifying this, check this when it turns green).

The following are required only for core and extension modules (they are welcomed, but not required, for unsupported modules):

- [x] There is an issue in [Jira](https://osgeo-org.atlassian.net/projects/GEOT) describing the bug/task/new feature (a notable exemptions is, changes not visible to end users). The ticket is for the GeoTools project, if the issue was found elsewhere it's a good practice to link to the origin ticket/issue.
- [x] The pull request contains changes related to a single objective. If multiple focuses cannot be avoided, each one is in its own commit and has a separate ticket describing it.
- [ ] PR for bug fixes and small new features are presented as a single commit
- [x] Commit message(s) must be in the form "[GEOT-XYZW] Title of the Jira ticket"
- [x] New unit tests have been added covering the changes
- [ ] This PR passes all existing unit tests (test results will be reported by Continuous Integration after opening this PR)
- [ ] This PR passes the [QA checks](https://docs.geotools.org/latest/developer/conventions/code/qa.html) (QA checks results will be reported by Continuous Integration after opening this PR)
- [ ] Documentation has been updated accordingly.

Submitting the PR does not require you to check all items, but by the time it gets merged, they should be either satisfied or not applicable.

### CODE REVIEW

File: modules/plugin/shapefile/src/main/java/org/geotools/data/shapefile/ShapefileIndexerBoundsHelper.java [hunk 1]
```diff
@@ -0,0 +1,325 @@
...
+    private static final Logger LOGGER = Logging.getLogger(ShapefileIndexerBoundsHelper.class);
+
+    /**
+     * A record count threshold past which bounds storage falls back to a temporary file.
+     *
+     * <p>Set to the equivalent of 1MB of heap memory, accounting for 32768 records (for a 32 byte
+     * serialized envelope size), or 65536 records if the shapefile geometry type is POINT (using 16
+     * bytes per envelope).
+     */
+    private static final int FALLBACK_TO_FILE_RECCOUNT_THRESHOLD = (1024 * 1024) / 32;
```
Comment (aaime) [inline modules/plugin/shapefile/src/main/java/org/geotools/data/shapefile/ShapefileIndexerBoundsHelper.java:83]:
```suggestion
    private static final int FALLBACK_TO_FILE_REC_COUNT_THRESHOLD = (1024 * 1024) / 32;
```

(at first I told myself... is that a typo? looks like "raccoon"!)

File: modules/plugin/shapefile/src/main/java/org/geotools/data/shapefile/ShapefileIndexerBoundsHelper.java [hunk 2]
```diff
@@ -0,0 +1,325 @@
...
+            reader.read(recNumber, env);
+        }
+
+        @Override
+        public void expand(int recNumber, Envelope env) throws IOException {
+            reader.expand(recNumber, env);
+        }
+
+        private void checkAvailableSpace(final Path path, final int fileSize)
+                throws IOException, FileSystemException {
```
Comment (aaime) [inline modules/plugin/shapefile/src/main/java/org/geotools/data/shapefile/ShapefileIndexerBoundsHelper.java:314]: FileSystemException is a subclass of IOException, so, redundant in the throw.

File: modules/plugin/shapefile/src/main/java/org/geotools/data/shapefile/ShapefileIndexerBoundsHelper.java [hunk 3]
```diff
@@ -0,0 +1,325 @@
...
+
+        if (numRecs <= fileRecCountThreshold) {
+            return new HeapBoundsReader(numRecs, pointBounds);
+        }
+        try {
+            return new FileBoundsReader(numRecs, pointBounds);
+        } catch (IOException e) {
+            LOGGER.log(
+                    Level.INFO,
+                    "Unable to create a temporary file backed bounds reader, falling back to slower ShapefileReader randome access",
```
Comment (aaime) [inline modules/plugin/shapefile/src/main/java/org/geotools/data/shapefile/ShapefileIndexerBoundsHelper.java:109]: Typo:

```suggestion
                    "Unable to create a temporary file backed bounds reader, falling back to slower ShapefileReader random access",
```

File: modules/plugin/shapefile/src/main/java/org/geotools/data/shapefile/ShapefileIndexerBoundsHelper.java [hunk 4]
```diff
@@ -0,0 +1,325 @@
...
+    static interface BoundsReader {
+
+        /** Records the bounds of the given shapefile record to be read after. */
+        void insert(int recNumber, Envelope env);
+
+        /** Initializes the provided envelope to the bounds of the requested shapefile record. */
+        void read(int recNumber, Envelope env) throws IOException;
+
+        /** Expands the provided envelope to cover the bounds of the shapefile record number. */
+        void expand(int recNumber, Envelope env) throws IOException;
```
Comment (aaime) [inline modules/plugin/shapefile/src/main/java/org/geotools/data/shapefile/ShapefileIndexerBoundsHelper.java:138]: Would recommend renaming to "expandEnvelope", this name seems like it would expand the reader somehow instead (at first I was looking for code that was growing the backing buffer/file, which did not make sense...)

**Review by aaime — CHANGES_REQUESTED** [review]
Nicely done, the speedup is indeed massive, the code is well documented and readable too. Added a couple of notes below.

File: modules/plugin/shapefile/src/main/java/org/geotools/data/shapefile/ShapefileIndexerBoundsHelper.java (hunk 1 above)
Comment (groldan) [reply modules/plugin/shapefile/src/main/java/org/geotools/data/shapefile/ShapefileIndexerBoundsHelper.java]: lol, done. Thanks for the review.

File: modules/plugin/shapefile/src/main/java/org/geotools/data/shapefile/ShapefileIndexerBoundsHelper.java (hunk 2 above)
Comment (groldan) [reply modules/plugin/shapefile/src/main/java/org/geotools/data/shapefile/ShapefileIndexerBoundsHelper.java]: Done.

File: modules/plugin/shapefile/src/main/java/org/geotools/data/shapefile/ShapefileIndexerBoundsHelper.java (hunk 3 above)
Comment (groldan) [reply modules/plugin/shapefile/src/main/java/org/geotools/data/shapefile/ShapefileIndexerBoundsHelper.java]: Fixed.

**Comment by groldan** [comment]
Amended to fix merge conflicts and respect the new PMD rules

**Review by aaime — APPROVED** [review]
